File: 0189-rotate-array/0189-rotate-array.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        N= len(nums)
        k = k%N
        
        def reverse(l,r):
            while r>=l:
                nums[l],nums[r]=nums[r],nums[l]
                l += 1
                r -= 1
        
        reverse(0,N-k-1)
        reverse(N-k,N-1)
        reverse(0,N-1)
        
        return
        

File: 0189-rotate-array/0189-rotate-array.java
class Solution {
    public void rotate(int[] nums, int k) {
        if (nums == null || nums.length == 0 || k == 0) {
            return;
        }

        int N = nums.length;
        k = k % N; // Handle cases where k is greater than N

        // Reverse the entire array
        reverse(nums, 0, N - 1);
        // Reverse the first k elements
        reverse(nums, 0, k - 1);
        // Reverse the rest of the array
        reverse(nums, k, N - 1);
    }

    // Helper method to reverse a portion of the array
    private void reverse(int[] nums, int l, int r) {
        while (l < r) {
            int temp = nums[l]; // Declare temp to hold the value
            nums[l] = nums[r];
            nums[r] = temp;
            l++;
            r--;
        }
    }
}
